ls1intum/Artemis
# Cancel scheduling of deleted programming exercises.

### Checklist

### Motivation and Context
Tasks are scheduled when a programming exercise is created / updated (e.g. locking/unlocking of repositories).
Currently, these tasks are not cancelled when the exercise is deleted and thus lead to an error once executed.

### Description
Cancel scheduled tasks for deleted programming exercises.

**Note:** Within the existing scheduling-logic, the exercise is retrieved from the database using the id. As we delete the exercise from the database and can not guarantee in which order the removal from the database and the cancellation of the scheduling is performed (as it might happen on different Artemis instances), we rely **only** on the id when cancelling the scheduling.

### Steps for Testing
**Note:** You need access to the logs of the test server or a local setup to test this PR.

1. Create an exam with 2 programming exercises and set the visible date e.g. 5 minutes in the future
2. Delete one of them
3. The log should show the message `Received schedule cancel for programming exercise`.
4. Once the visible date is reached, the log should show the message `Invoking (scheduled) task 'add write permissions to all student repositories' for programming exercise with id` only for the not-deleted exercise
5. No errors should be present in the logs

**These testing steps can be performed even without access to the logs:**
1. Create an exam with a programming exercise and register students
2. Make sure that the repositories are correctly unlocked once the visible date is reached
3. Make sure that the repositories are correctly locked once the end date is reached

### Test Coverage
Unchanged
